tfjs-backend-webgpu/src/resize_nearest_neighbor_backprop_webgpu.ts

```diff
@@ -0,0 +1,113 @@
+ * limitations under the License.
+ * =============================================================================
+ */
+
+import {getMainHeaderString as main, WebGPUProgram} from './webgpu_program';
+import {computeDispatch, flatDispatchLayout} from './webgpu_util';
+
+export class ResizeNearestNeigborBackpropProgram implements WebGPUProgram {
+  outputShape: number[];
+  shaderKey: string;
+  dispatchLayout: {x: number[]};
+  dispatch: [number, number, number];
+  variableNames = ['dy'];
+  uniforms =
+      `effectiveXSize : vec2<i32>, effectiveYSize : vec2<i32>, invHeightScale : f32, invWidthScale : f32,
+       winHeight : i32, winWidth : i32, yHeight : i32, yWidth : i32, xHeight : i32, xWidth : i32,`;
+  workgroupSize: [number, number, number] = [64, 1, 1];
+  alignCorners: boolean;
+  size = true;
+
+  constructor(
+      inputShape: [number, number, number, number], alignCorners: boolean) {
+    this.outputShape = inputShape;
+
+    this.dispatchLayout = flatDispatchLayout(this.outputShape);
+    this.dispatch = computeDispatch(
+        this.dispatchLayout, this.outputShape, this.workgroupSize);
+
+    this.alignCorners = alignCorners;
+    this.shaderKey = `resizeNearestNeigborBackprop_${alignCorners}`;
+  }
+
+  getUserCode(): string {
+    const userCode = `
+      ${main('index')} {
+        if (index < uniforms.size) {
+          let coords = getOutputCoords();
+          let b = coords[0];
+          let d = coords[3];
+          let r = coords[1];
+          let c = coords[2];
+
+          var accumulator = 0.0;
+
+          // Compute bounds for where in dy we will look
+          let startRLerp = floor(f32(r) * uniforms.invHeightScale);
+          let startDyR = i32(floor(startRLerp - f32(uniforms.winHeight / 2)));
+
+          let startCLerp = floor(f32(c) * uniforms.invWidthScale);
+          let startDyC = i32(floor(startCLerp - f32(uniforms.winWidth / 2)));
+
+          // Loop over dy
+          for (var dyROffset = 0; dyROffset < uniforms.winHeight; dyROffset++) {
+            let dyR = startDyR + dyROffset;
+
+            // Guard against the window exceeding the bounds of dy
+            if (dyR < 0 || dyR >= uniforms.yHeight) {
+              continue;
+            }
+
+            for (var dyCOffset = 0; dyCOffset < uniforms.winWidth; dyCOffset++) {
+              let dyC = startDyC + dyCOffset;
+
+              // Guard against the window exceeding the bounds of dy
+              if (dyC < 0 || dyC >= uniforms.yWidth) {
+                continue;
+              }
+
+              let sourceFracRow = f32(uniforms.effectiveXSize[0]) *
+                  (f32(dyR) / f32(uniforms.effectiveYSize[0]));
+
+              let sourceFracCol = f32(uniforms.effectiveXSize[1]) *
+                  (f32(dyC) / f32(uniforms.effectiveYSize[1]));
+
+              let sourceNearestRow =
+                  i32(min(f32(uniforms.xHeight - 1),
+                  ${
+        this.alignCorners ? 'floor(sourceFracRow + 0.5)' :
```

Comment:
From WGSL spec, we could know that halfway cases are rounded to the nearest even, for example, `round(2.5) = 3` and `round(3.5) =3`, the behavior is not expected, we expect that `round(2.5) = 2 && round(3.5) = 3` or `round(2.5) = 3 && round(3.5) = 4`.
From GLSL spec, The fraction 0.5 will round in a direction chosen by the implementation, presumably the direction that is fastest, https://registry.khronos.org/OpenGL-Refpages/gl4/html/round.xhtml, I am not sure whether tfjs webgl backend will get different results on different machines when using `round` here.
From https://github.com/tensorflow/tfjs/blob/b114ff291a58bb211c330e1c84c5220260b2a310/tfjs-backend-webgpu/src/resize_nearest_neighbor_webgpu.ts#L81, ` ResizeNearestNeighborProgram` use `+0.5` before rounding the value, so we also use `+0.5`, not `-0.5` here.

The other comment was addressed, thank you.

Reply:
WebGPU has some discussion at https://github.com/gpuweb/gpuweb/issues/1381. Can we change the WebGL impl to use roundEven instead, so that its behavior is the same on all machines and same as WebGPU?

Reply:
Per sync with Google, we'll unify the solution to this, floor(x+0.5), for WebGL and WebGPU, so that all the backends, including CPU, have same behavior.